core/db_classes/EE_Registration.class.php

```diff
@@ -1925,13 +2294,30 @@ public function ticket_line_item()
     /**
      * Soft Deletes this model object.
      *
+     * @param string $source function name that called this method
      * @return boolean | int
-     * @throws RuntimeException
+     * @throws DomainException
      * @throws EE_Error
+     * @throws EntityNotFoundException
+     * @throws InvalidArgumentException
+     * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
+     * @throws ReflectionException
+     * @throws RuntimeException
+     * @throws UnexpectedEntityException
      */
-    public function delete()
+    public function delete($source = 'unknown')
     {
         if ($this->update_extra_meta(EE_Registration::PRE_TRASH_REG_STATUS_KEY, $this->status_ID()) === true) {
+            $current_user = wp_get_current_user();
+            $this->add_extra_meta(
+                EE_Registration::EXTRA_META_KEY_REG_TRASHED,
+                array(
+                    'trashed-by' => $current_user->ID ? $current_user->display_name : 'unauthed user',
+                    'timestamp'  => time(),
+                    'source'     => $source,
+                )
+            );
```

Comment:
extra logging

Reply:
I'm not sure I like adding a parameter to the method for the sole purpose of logging.  I'd prefer if we implemented actions that logging can hook into.  For example, anyplace we _call_ this method, you can add a `do_action()` and then add a do_action here in this method.

So for example:

`EE_Registration:set_deleted()` could have:

```php
/**
 * Sets deleted
 *
 * @param boolean $deleted
 * @return bool
 * @throws EE_Error
 * @throws RuntimeException
 */
public function set_deleted($deleted)
{
    do_action('AHEE__EE_Registration__set_deleted__before', $deleted, $this);
    if ($deleted) {
        $this->delete();
    } else {
        $this->restore();
    }
    do_action('AHEE__EE_Registration__set_deleted_after', $deleted, $this);
}
```

and `EE_Registration::deleted` could have:

```php
/**
 * Soft Deletes this model object.
 *
 * @return boolean | int
 * @throws RuntimeException
 * @throws EE_Error
 */
public function delete()
{
    do_action('AHEE__EE_Registration__delete__before', $this);
    if ($this->update_extra_meta(EE_Registration::PRE_TRASH_REG_STATUS_KEY, $this->status_ID()) === true) {
        $this->set_status(EEM_Registration::status_id_cancelled);
    }
    return parent::delete();
}
```

Then consuming code could do something like this:

```php
function setDeletedLogging($deleted, $registration) {
    $method = 'EE_Registration::set_deleted';
    $logger = function($registration) use($method) {
        // do the extra_meta logging in here using $method etc.
    };
    // perform logging
    add_action('AHEE__EE_Registration__delete__before', $logger);
    // remove logging
    add_action('AHEE__EE_Registration__set_deleted_after', function($deleted, $registration) use($logger) {
        remove_action('AHEE__EE_Registration__set_deleted__before', 'setDeletedLogging', 10, 2 );
        remove_action('AHEE__EE_Registration_delete_before', $logger);
    } );
}

// set the logging
add_action('AHEE__EE_Registration__set_deleted__before', 'setDeletedLogging', 10, 2 );
```

I realize my suggestion is more verbose but it allows us to keep the core code cleaner and also exposes the actions for other potential future needs as well.  Plus we're not unnecessarily doing this logging on ALL sites out in the wild (which some people could be upset about because of the increase in database size).